### ISSUE TITLE
Trying to update a document not in collection

### ISSUE THREAD

**Issue by Alex-Werner** [techfort/LokiJS#405]
Hello ! 
Thank a lot for the work on LokiJS. I'm actually using it at work but I have two issue. 

First one is when I benchmark LokiJS, when I start on fresh database file, insert aren't a problem (it takes 6ms to insert). 
But then after 6000+ insertions (users object), It takes approx 850ms~ So, on that, I really need to try to check a lot more with memdump and stuff like that to see where did I goes wrong, but can you confirm that if I have this kind of problem, that's an implementation problem in my side ? 
Additionnally, I need to handle 1 000 000 differents users in the database, is it possible ? I couldn't go that far (because of the 1second insert time after 6K+) 

Second one is something I don't know how to solve because I don't know what causing it. 
I get a User inserted, I modify one property, and then, i update, and it throw me a "uncaught error: Trying to update a document not in collection" 

![loki_err_censored](https://cloud.githubusercontent.com/assets/5849920/14737191/19cf5ace-087b-11e6-89f1-5d776fc9de80.png)

EDIT : I forgot to specify. It's the second time it's happens. The first time I didn't expect this so I didn't have any idea on what could causing it. 
The second time was this morning when I wake up my computer, it was in a low-power standby mode.
When I erase my DB, then It works perfectly. I don't know yet if it's linked to the low-power mode. 

Thank you a lot for your answers ! 

Have a good day !

**Comment by obeliskos** [comment]
If its -consistently- above 10ms or so after 6000 docs we might want to have a look at that.  Otherwise an occasional 850ms spike could be acceptably attributed to javascript memory management / housekeeping.  Maybe you could post a plunkr or gist for loki sandbox which demonstrates.

Your other issue example seems like you should be calling findOne() instead of find().  Your find() call is probably returning an array of 1 document and you are attempting to set a property on the array, not the first document in that array.

**Comment by Alex-Werner** [comment]
Hi obeliskos, thank you for your answers. I've updated my first post adding useful informations that I should said earlier. 

Here are the benchmark data. 

![15_avril_2016-benchmark-0-users](https://cloud.githubusercontent.com/assets/5849920/14739582/ca916fb6-0888-11e6-908f-77a6a7b51a03.png)
This is at start. I set db and let's go to the add party. It displays the number of insertions being done. We can see it dropping. 

![15_avril_2016-benchmark-6000-users](https://cloud.githubusercontent.com/assets/5849920/14739584/cc2c9d78-0888-11e6-8498-87358a4bc587.png)
Here is the part that makes me feels I have a real issue on my side, but I wanted to ask to be sure (I've searched on the issues already closed but didn't find yet but well, had to try)

![15_avril_2016-benchmark-1](https://cloud.githubusercontent.com/assets/5849920/14739592/d68b3efa-0888-11e6-9ec5-b24126de7d9b.png)
Here there is 4 graphs. Top-left is users added. Top-right is ops (red is add Users), get Users or Generate (faker.js/chance.js) Users are at 0 because there are too fast and makes my red chart unreadable so they are unactivate. 
Bottom-left is the DB size in Mib. Bottom-right is Mem Usage: green : heapTotal, red:heapUsed, blue: RSS.  

![15_avril_2016-benchmark-2](https://cloud.githubusercontent.com/assets/5849920/14739686/864de93c-0889-11e6-8883-9cedd7f2f03b.png)

![15_avril_2016-benchmark-3](https://cloud.githubusercontent.com/assets/5849920/14739594/d9bfc21c-0888-11e6-9e4f-90c9a2364fb6.png)

![15_avril_2016-benchmark-4](https://cloud.githubusercontent.com/assets/5849920/14739684/80f621c0-0889-11e6-9647-5b91273a838c.png)
![15_avril_2016-benchmark-5](https://cloud.githubusercontent.com/assets/5849920/14739691/8cef733c-0889-11e6-8a45-0f5ae5cd4d44.png)

![15_avril_2016-benchmark-6](https://cloud.githubusercontent.com/assets/5849920/14739596/daf2a71c-0888-11e6-97f6-73797e9f5944.png)


--- 
I have plan to do deeper search, deeper analysis & better benchmarking/reporting to try to find what is causing that. 
I would be very pleased to know every advice you could have or any particular stuff you would me to monitor or else. 

----

About the array set of property : Damn indeed, I didn't see that on the first place :O 
But after deleting my db (restarting) it didn't throw me any errors on that part. Weird,
I will buy a better brains and learn to do proper exemples and check that again.

But on my initial method, the returns of my array is normal and handled, it works on 95% of time, it just failed to times. But when I delete my DB and recreate it, it works again.

---- 

Thank you a lot !

**Comment by obeliskos** [comment]
I won't have a chance to look over this until this weekend but this looks very helpful.  Yes as much diagnostics as you can offer would be great and we could add or update our benchmarks to help find and fix any problems.

In general I should think the only things that might be relevant are indices (binary or exact), cloning, and DynamicViews (with inefficient pipelines).  If you are doing none of those then there is an even more mysterious cause to your degraded performance.  

Post any of your results or further analysis and I will get back over the weekend.  Thanks!

**Comment by obeliskos** [comment]
Not really able to reproduce this.  My test of simple insertion benchmark is inserting hundreds of thousands of documents with little performance falloff.

[https://rawgit.com/techfort/LokiJS/master/examples/sandbox/LokiSandbox.htm#rawgist=https://gist.githubusercontent.com/obeliskos/c5be9535a3a2d94087ca341794b24256/raw/ac6e65b6963feb721a1a9cf05cca33682150aa82/lokiInsertBenchmark](Here is a loki sandbox gist)

Plug in higher or lower values but the net ops/s are roughly the same.  

Feel free to provide more details on how big your objects are are any database and collection options you are using.  If you had autosave with very low interval (under a second or two) it might do that, or collection cloning, indexes, etc.

**Comment by Alex-Werner** [comment]
Hi ! i'm sorry I took so long to reply. 
But I had a huge amount of work to do, and I struggled with this issue since. 
After more deep search on this today, I found that what was making me this pain in the ass was the Collection.prototype.get method. 
See over here : 



![nouserbreakpoint_](https://cloud.githubusercontent.com/assets/5849920/15245564/9a5ed31c-1907-11e6-9e84-91bdbcb1c954.png)


So I gone deeper, and found that Get was a binary search, but it do it on an un-sorted array. See this : 

![debug_](https://cloud.githubusercontent.com/assets/5849920/15245566/a05687ce-1907-11e6-999b-d2ee08e84ef6.png)

So I try, and if you have this : 
data = [12,1,4,9,3,8,11,5,7,6,2,10,13]
Collection.prototype.get(1,true) will return null. 
Now, I overwrited Collection.prototype.get with a way slower for method. I will give it a try on the next days.

My question being this one : What did I do to make this array (data and idIndex) unsorted ? I just used : findOne, find and filter to deal with my data. Can filter do that ? 

    databases._visits=createAndGetCollection('visits');
    databases._visits.filter(function(visit){ //doStuffWithVisit });

**Comment by obeliskos** [comment]
Hah, that's confusing to follow.  Not sure how your idIndex got corrupted.

Javascript filter should be safe unless you are referencing outside the current object into some global loki db instance.

If your idIndex is just out of sequence, you might try doing something like this : 
```javascript
db.getCollection('visits').idIndex.sort();
```
That would at least fix the get() binary search's ability to find the documents.

Maybe post a gist or plunkr if you can recreate on fresh database.

**Comment by Alex-Werner** [comment]
I will do that. 
Actually I mostly use find, findOne, update and insert only. 
I just used LokiJS in order to store my documents, but as the project weight a lot I have trouble to know where this cames from (as it's very hard to reproduce, in the last 23 days - having the server running 24h/24 we had seen this problem happens may be 5 or 6 times only (as it's a dev server we just cleaned the DB when it happens). 
 
For now as i'm late in my duty, I have to quickly switch on MongoDB (we had to go live one weeks ago :x), so I can't give more information  before at least wednesday.

I will then have more time to try to isolate a test case that make it happens. I will then remove all the sensitive information and try to give that in a gist. 

Sorry for making it too long to gives more code information, but given the hard-to-reproduce even in my sides and the current NDA i'm on this project I can't just copy-paste easily the code. 
But I will try to determinate as quick as I can where does this problem cames from.
